src/cpp/session/modules/SessionThemes.cpp

```diff
@@ -0,0 +1,176 @@
+/*
+ * SessionThemes.cpp
+ *
+ * Copyright (C) 2018 by RStudio, Inc.
+ *
+ * Unless you have received this program directly from RStudio pursuant
+ * to the terms of a commercial license agreement with RStudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+
+#include "SessionThemes.hpp"
+
+#include <boost/bind.hpp>
+#include <boost/filesystem.hpp>
+
+#include <core/Error.hpp>
+#include <core/Exec.hpp>
+#include <core/json/JsonRpc.hpp>
+
+#include <session/SessionModuleContext.hpp>
+
+#include <r/RRoutines.hpp>
+#include <r/RSexp.hpp>
+
+#include <map>
+#include <regex>
+#include <string>
+
+using namespace rstudio::core;
+
+namespace rstudio {
+namespace session {
+namespace modules {
+namespace themes {
+
+namespace {
+
+// A map from the name of the theme to the location of the file and a boolean representing
+// whether or not the theme is dark.
+typedef std::map<std::string, std::pair<std::string, bool> > ThemeMap;
+void getThemesInLocation(const boost::filesystem::path& location, ThemeMap& themeMap)
```

Comment:
I think the main reason is to ensure exception safety, since many of the Boost filesystem APIs can throw exceptions (mostly in rare cases where e.g. the disk gets unmounted while we're doing something filesystem-y on it). We make sure to catch and log those and then return a 'sane' value when that happens.